contracts/staking/GoodCompoundStaking.sol

```diff
@@ -1,94 +1,147 @@
+	 */
+	constructor(
+		address _token,
+		address _iToken,
+		uint256 _blockInterval,
+		NameService _ns,
+		string memory _tokenName,
+		string memory _tokenSymbol,
+		uint64 _maxRewardThreshold,
+        address _tokenUsdOracle
+	)
+		SimpleStaking(
+			_token,
+			_iToken,
+			_blockInterval,
+			_ns,
+			_tokenName,
+			_tokenSymbol,
+			_maxRewardThreshold
+		)
+	{
+        tokenUsdOracle = _tokenUsdOracle;
     }
+    // Address of the TOKEN/USD oracle from chainlink
+	address public tokenUsdOracle;
+	/**
+	 * @dev stake some Token
+	 * @param _amount of Token to stake
+	 */
+	function mintInterestToken(uint256 _amount) internal override {
+		cERC20 cToken = cERC20(address(iToken));
+		uint256 res = cToken.mint(_amount);
+
+	    require(res == 0, "Minting cToken failed, funds returned");
+	
+	}
+
+	/**
+	 * @dev redeem Token from compound
+	 * @param _amount of token to redeem in Token
+	 */
+	function redeem(uint256 _amount) internal override {
+		cERC20 cToken = cERC20(address(iToken));
+		require(cToken.redeemUnderlying(_amount) == 0, "Failed to redeem cToken");
+	}
 
     /**
-     * @dev stake some DAI
-     * @param _amount of dai to stake
+     * @dev Function to redeem cToken for DAI
+     * @dev _amount of token in iToken
+     * @return return address of the DAI and amount of the DAI
      */
-    function mintInterestToken(uint256 _amount) internal override{
-        
-        cERC20 cToken = cERC20(address(iToken));
-        uint res = cToken.mint(_amount);
-
-        if (
-            res > 0
-        ) //cDAI returns >0 if error happened while minting. make sure no errors, if error return DAI funds
-        {
-            require(res == 0, "Minting cDai failed, funds returned");
-        }
-
-    }
+	function redeemUnderlying(uint256 _amount)
+		internal
+		override
+		returns (address, uint256)
+	{
 
-    /**
-     * @dev redeem DAI from compound 
-     * @param _amount of dai to redeem
-     */
-    function redeem(uint256 _amount) internal override{
+        uint256 tokenBalance = token.balanceOf(address(this));
         cERC20 cToken = cERC20(address(iToken));
-        require(cToken.redeemUnderlying(_amount) == 0, "Failed to redeem cDai");
-
-    }
-
-    /**
-     * @dev returns Dai to cDai Exchange rate.
-     */
-    function exchangeRate() internal view override returns(uint) {
-        cERC20 cToken = cERC20(address(iToken));
-        return cToken.exchangeRateStored();
-
-    }
-
-    /**
-     * @dev returns decimals of token.
-     */
-    function tokenDecimal() internal view override returns(uint) {
-        ERC20 token = ERC20(address(token));
-        return uint(token.decimals());
-    }
-
-    /**
-     * @dev returns decimals of interest token.
+        require(cToken.redeem(_amount) == 0, "Failed to redeem cToken");
+        uint256 redeemedAmount = token.balanceOf(address(this)) - tokenBalance;
+        if(address(iToken) == nameService.getAddress("CDAI")){
```

Comment:
we can optimize this
just return cDAI without redeeming to DAI.
see FundManager comment

Reply:
Changed